In [251]:
# List of RTD function calls for AAPL
rtd_calls_list = [
    '=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")',
    '=RTD("tos.rtd", , "LAST", "AAPL")',
    '=RTD("tos.rtd", , "LAST_SIZE", "AAPL")',
    '=RTD("tos.rtd", , "BID_SIZE", "AAPL")',
    '=RTD("tos.rtd", , "BID", "AAPL")',
    '=RTD("tos.rtd", , "ASK", "AAPL")',
    '=RTD("tos.rtd", , "ASK_SIZE", "AAPL")',
    '=RTD("tos.rtd", , "NET_CHANGE", "AAPL")',
    '=RTD("tos.rtd", , "PERCENT_CHANGE", "AAPL")',
    '=RTD("tos.rtd", , "MARK_PERCENT_CHANGE", "AAPL")',
    '=RTD("tos.rtd", , "PUT_CALL_RATIO", "AAPL")',
    '=RTD("tos.rtd", , "CALL_VOLUME_INDEX", "AAPL")',
    '=RTD("tos.rtd", , "PUT_VOLUME_INDEX", "AAPL")',
    '=RTD("tos.rtd", , "OPTION_VOLUME_INDEX", "AAPL")',
    '=RTD("tos.rtd", , "BACK_EX_MOVE", "AAPL")',
    '=RTD("tos.rtd", , "FRONT_EX_MOVE", "AAPL")',
    '=RTD("tos.rtd", , "EX_MOVE_DIFF", "AAPL")',
    '=RTD("tos.rtd", , "IMPL_VOL", "AAPL")',
    '=RTD("tos.rtd", , "Spreads", "AAPL")',
    '=RTD("tos.rtd", , "MRKT_MKR_MOVE", "AAPL")',
    '=RTD("tos.rtd", , "MarketSentiment", "AAPL")',
    '=RTD("tos.rtd", , "VOLUME", "AAPL")'
]

# Each RTD call is already a string, so we can just print them out or use them as needed
rtd_calls_list


['=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")',
 '=RTD("tos.rtd", , "LAST", "AAPL")',
 '=RTD("tos.rtd", , "LAST_SIZE", "AAPL")',
 '=RTD("tos.rtd", , "BID_SIZE", "AAPL")',
 '=RTD("tos.rtd", , "BID", "AAPL")',
 '=RTD("tos.rtd", , "ASK", "AAPL")',
 '=RTD("tos.rtd", , "ASK_SIZE", "AAPL")',
 '=RTD("tos.rtd", , "NET_CHANGE", "AAPL")',
 '=RTD("tos.rtd", , "PERCENT_CHANGE", "AAPL")',
 '=RTD("tos.rtd", , "MARK_PERCENT_CHANGE", "AAPL")',
 '=RTD("tos.rtd", , "PUT_CALL_RATIO", "AAPL")',
 '=RTD("tos.rtd", , "CALL_VOLUME_INDEX", "AAPL")',
 '=RTD("tos.rtd", , "PUT_VOLUME_INDEX", "AAPL")',
 '=RTD("tos.rtd", , "OPTION_VOLUME_INDEX", "AAPL")',
 '=RTD("tos.rtd", , "BACK_EX_MOVE", "AAPL")',
 '=RTD("tos.rtd", , "FRONT_EX_MOVE", "AAPL")',
 '=RTD("tos.rtd", , "EX_MOVE_DIFF", "AAPL")',
 '=RTD("tos.rtd", , "IMPL_VOL", "AAPL")',
 '=RTD("tos.rtd", , "Spreads", "AAPL")',
 '=RTD("tos.rtd", , "MRKT_MKR_MOVE", "AAPL")',
 '=RTD("tos.rtd", , "MarketSentiment", "AAPL")',
 '=RTD("tos.rtd", , "VOLUME", "AAPL")']

In [252]:
new_list = [
    '=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""LAST"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""LAST_SIZE"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""BID_SIZE"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""BID"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""ASK"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""ASK_SIZE"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""NET_CHANGE"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""PERCENT_CHANGE"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""MARK_PERCENT_CHANGE"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""PUT_CALL_RATIO"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""CALL_VOLUME_INDEX"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""PUT_VOLUME_INDEX"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""OPTION_VOLUME_INDEX"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""BACK_EX_MOVE"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""FRONT_EX_MOVE"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""EX_MOVE_DIFF"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""IMPL_VOL"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""Spreads"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""MRKT_MKR_MOVE"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""MarketSentiment"", ""AAPL"")',
    '=RTD(""tos.rtd"", , ""VOLUME"", ""AAPL"")'
]

new_list

['=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""LAST"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""LAST_SIZE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""BID_SIZE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""BID"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""ASK"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""ASK_SIZE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""NET_CHANGE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""PERCENT_CHANGE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""MARK_PERCENT_CHANGE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""PUT_CALL_RATIO"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""CALL_VOLUME_INDEX"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""PUT_VOLUME_INDEX"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""OPTION_VOLUME_INDEX"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""BACK_EX_MOVE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""FRONT_EX_MOVE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""EX_MOVE_DIFF"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""IMPL_VOL"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""Spreads"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""MRKT_MKR_MOVE"",

In [253]:
# You can add any ticker you want here. We will pull in the list of NQ+ES MAKE THIS FRONT END
fake_spy = ["AAPL", "AMZN", "QQQ"]


# Iterate through the calls. 
# AAPL needs to become 'x'
# We change AAPL in each row with 'x'
# We have a list of stocks that we want to update 'x' to be
# First, x = AAPL, then x = AMZN, then x = QQQ, for all tickers in 'fake_spy'

rtd_call_per_ticker = []

for Ticker in fake_spy:
    fake_rtd_calls = [
        '=RTD("tos.rtd", , "MARK", "' + Ticker + '")',
        '=RTD("tos.rtd", , "LAST", "' + Ticker + '")',
        '=RTD("tos.rtd", , "LAST_SIZE", "' + Ticker + '")',
        '=RTD("tos.rtd", , "BID_SIZE", "' + Ticker + '")',
        '=RTD("tos.rtd", , "BID", "' + Ticker + '")',
        '=RTD("tos.rtd", , "ASK", "' + Ticker + '")',
        '=RTD("tos.rtd", , "ASK_SIZE", "' + Ticker + '")',
        '=RTD("tos.rtd", , "NET_CHANGE", "' + Ticker + '")',
        '=RTD("tos.rtd", , "PERCENT_CHANGE", "' + Ticker + '")',
        '=RTD("tos.rtd", , "MARK_PERCENT_CHANGE", "' + Ticker + '")',
        '=RTD("tos.rtd", , "PUT_CALL_RATIO", "' + Ticker + '")',
        '=RTD("tos.rtd", , "CALL_VOLUME_INDEX", "' + Ticker + '")',
        '=RTD("tos.rtd", , "PUT_VOLUME_INDEX", "' + Ticker + '")',
        '=RTD("tos.rtd", , "OPTION_VOLUME_INDEX", "' + Ticker + '")',
        '=RTD("tos.rtd", , "BACK_EX_MOVE", "' + Ticker + '")',
        '=RTD("tos.rtd", , "FRONT_EX_MOvE", "' + Ticker + '")',
        '=RTD("tos.rtd", , "EX_MOVE_DIFF", "' + Ticker + '")',
        '=RTD("tos.rtd", , "IMPL_VOL", "' + Ticker + '")',
        '=RTD("tos.rtd", , "Spreads", "' + Ticker + '")',
        '=RTD("tos.rtd", , "MRKT_MKR_MOVE", "' + Ticker + '")',
        '=RTD("tos.rtd", , "MarketSentiment", "' + Ticker + '")',
        '=RTD("tos.rtd", , "VOLUME", "' + Ticker + '")',
    ]
    rtd_call_per_ticker.append(fake_rtd_calls)

rtd_call_per_ticker

[['=RTD("tos.rtd", , "MARK", "AAPL")',
  '=RTD("tos.rtd", , "LAST", "AAPL")',
  '=RTD("tos.rtd", , "LAST_SIZE", "AAPL")',
  '=RTD("tos.rtd", , "BID_SIZE", "AAPL")',
  '=RTD("tos.rtd", , "BID", "AAPL")',
  '=RTD("tos.rtd", , "ASK", "AAPL")',
  '=RTD("tos.rtd", , "ASK_SIZE", "AAPL")',
  '=RTD("tos.rtd", , "NET_CHANGE", "AAPL")',
  '=RTD("tos.rtd", , "PERCENT_CHANGE", "AAPL")',
  '=RTD("tos.rtd", , "MARK_PERCENT_CHANGE", "AAPL")',
  '=RTD("tos.rtd", , "PUT_CALL_RATIO", "AAPL")',
  '=RTD("tos.rtd", , "CALL_VOLUME_INDEX", "AAPL")',
  '=RTD("tos.rtd", , "PUT_VOLUME_INDEX", "AAPL")',
  '=RTD("tos.rtd", , "OPTION_VOLUME_INDEX", "AAPL")',
  '=RTD("tos.rtd", , "BACK_EX_MOVE", "AAPL")',
  '=RTD("tos.rtd", , "FRONT_EX_MOvE", "AAPL")',
  '=RTD("tos.rtd", , "EX_MOVE_DIFF", "AAPL")',
  '=RTD("tos.rtd", , "IMPL_VOL", "AAPL")',
  '=RTD("tos.rtd", , "Spreads", "AAPL")',
  '=RTD("tos.rtd", , "MRKT_MKR_MOVE", "AAPL")',
  '=RTD("tos.rtd", , "MarketSentiment", "AAPL")',
  '=RTD("tos.rtd", , "VOLUME", "AAPL"

Now we are going to do the same with ALL stocks in both the NQ and ES

In [254]:
# Import the required libraries
import numpy as np
import pandas as pd

In [255]:
# Read csv, take a look
nq_df = pd.read_csv(("../algotrader2/resources/nq_stocks.csv"))
nq_df.head()

Symbol           Description      Category2  Category3  \
0   AAPL             Apple Inc  Common stocks  Large cap   
1   MSFT        Microsoft Corp  Common stocks  Large cap   
2   GOOG  Alphabet Inc Class C  Common stocks  Large cap   
3  GOOGL  Alphabet Inc Class A  Common stocks  Large cap   
4   AMZN       Amazon.Com Inc.  Common stocks  Large cap   

              GICS Sector          Market cap Dividend yield Country   Action  
0  Information Technology  $2,514,581,877,780          0.58%     USA  Analyze  
1  Information Technology  $2,066,846,488,973          0.93%     USA  Analyze  
2  Communication Services  $1,263,113,280,000          0.00%     USA  Analyze  
3  Communication Services  $1,263,113,280,000          0.00%     USA  Analyze  
4  Consumer Discretionary  $1,011,507,010,613          0.00%     USA  Analyze

In [256]:
# Do the same for the snp 500 symbols
snp_df = pd.read_csv(("../algotrader2/resources/snp_stocks.csv"))
snp_df.head()

Symbol     Security             GICS Sector               GICS Sub-Industry  \
0    MMM           3M             Industrials        Industrial Conglomerates   
1    AOS  A. O. Smith             Industrials               Building Products   
2    ABT       Abbott             Health Care           Health Care Equipment   
3   ABBV       AbbVie             Health Care                   Biotechnology   
4    ACN    Accenture  Information Technology  IT Consulting & Other Services   

     Headquarters Location  Date added      CIK      Founded  
0    Saint Paul, Minnesota  1957-03-04    66740         1902  
1     Milwaukee, Wisconsin  2017-07-26    91142         1916  
2  North Chicago, Illinois  1957-03-04     1800         1888  
3  North Chicago, Illinois  2012-12-31  1551152  2013 (1888)  
4          Dublin, Ireland  2011-07-06  1467373         1989

In [257]:
# # Save the symbols from each df
# nq_symbols_df = nq_df["Symbol"]
# snp_symbols_df = snp_df["Symbol"]

# Save symbols from each df and put them together
snp_nq_symbols_df = pd.concat([nq_df["Symbol"], snp_df["Symbol"]], axis=0)

# Drop duplicates
snp_nq_symbols_df.drop_duplicates(inplace=True)

# View
snp_nq_symbols_df

0       AAPL
1       MSFT
2       GOOG
3      GOOGL
4       AMZN
       ...  
498      YUM
499     ZBRA
500      ZBH
501     ZION
502      ZTS
Name: Symbol, Length: 521, dtype: object

As we can see, the index is wrong. Somewhere we have duplicate indices, since we end at 502 but apparently have 521. We may come back to this, but I don't think it matters.

In [258]:
# Convert symbols to a list so that we can feed it into our rtd_call function

nq_snp_symbols_list = snp_nq_symbols_df.tolist()

print(f"Number of unique stocks: {len(nq_snp_symbols_list)}")
nq_snp_symbols_list[:10]


Number of unique stocks: 521


['AAPL',
 'MSFT',
 'GOOG',
 'GOOGL',
 'AMZN',
 'NVDA',
 'TSLA',
 'META',
 'ASML',
 'AVGO']

In [259]:
fake_list = []


for Ticker in fake_spy:
    fake_calls = [
        '=RTD(""tos.rtd"", , ""MARK"", ""' + Ticker + '"")',
        '=RTD(""tos.rtd"", , ""LAST"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""LAST_SIZE"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""BID_SIZE"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""BID"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""ASK"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""ASK_SIZE"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""NET_CHANGE"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""PERCENT_CHANGE"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""MARK_PERCENT_CHANGE"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""PUT_CALL_RATIO"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""CALL_VOLUME_INDEX"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""PUT_VOLUME_INDEX"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""OPTION_VOLUME_INDEX"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""BACK_EX_MOVE"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""FRONT_EX_MOvE"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""EX_MOVE_DIFF"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""IMPL_VOL"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""Spreads"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""MRKT_MKR_MOVE"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""MarketSentiment"", ""Ticker"")',
        '=RTD(""tos.rtd"", , ""VOLUME"", ""Ticker"")'
    ]
    fake_list.append(fake_calls)
    

fake_list

[['=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""LAST"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""LAST_SIZE"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""BID_SIZE"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""BID"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""ASK"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""ASK_SIZE"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""NET_CHANGE"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""PERCENT_CHANGE"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""MARK_PERCENT_CHANGE"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""PUT_CALL_RATIO"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""CALL_VOLUME_INDEX"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""PUT_VOLUME_INDEX"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""OPTION_VOLUME_INDEX"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""BACK_EX_MOVE"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""FRONT_EX_MOvE"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""EX_MOVE_DIFF"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""IMPL_VOL"", ""Ticker"")',
  '=RTD(""tos.rtd"", , ""Spreads"",

In [260]:
# Skeleton list of lists. Each element is a list of all of calls we want to make, given each stock
rtd_call_per_ticker = []

# Iterate through the calls. 
# AAPL needs to become 'x'
# We change AAPL in each row with 'x'
# We have a list of stocks that we want to update 'x' to be
# First, x = AAPL, then x = AMZN, then x = QQQ, for all tickers in nq_snp_symbols_list


# Iterate through tickers in our combined nq/snp list.
# Keep adding those lists to our list of lists.
# Fixed the double quotation problem with help from ChatGPT
for Ticker in nq_snp_symbols_list:
    real_call_list = [
    '=RTD(""tos.rtd"", , ""MARK"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""LAST"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""LAST_SIZE"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""BID_SIZE"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""BID"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""ASK"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""ASK_SIZE"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""NET_CHANGE"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""PERCENT_CHANGE"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""MARK_PERCENT_CHANGE"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""PUT_CALL_RATIO"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""CALL_VOLUME_INDEX"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""PUT_VOLUME_INDEX"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""OPTION_VOLUME_INDEX"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""BACK_EX_MOVE"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""FRONT_EX_MOvE"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""EX_MOVE_DIFF"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""IMPL_VOL"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""Spreads"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""MRKT_MKR_MOVE"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""MarketSentiment"", ""' + Ticker + '"")',
    '=RTD(""tos.rtd"", , ""VOLUME"", ""' + Ticker + '"")',
    ]
    rtd_call_per_ticker.append(real_call_list)
    

In [261]:
rtd_call_per_ticker

[['=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""LAST"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""LAST_SIZE"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""BID_SIZE"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""BID"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""ASK"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""ASK_SIZE"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""NET_CHANGE"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""PERCENT_CHANGE"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""MARK_PERCENT_CHANGE"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""PUT_CALL_RATIO"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""CALL_VOLUME_INDEX"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""PUT_VOLUME_INDEX"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""OPTION_VOLUME_INDEX"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""BACK_EX_MOVE"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""FRONT_EX_MOvE"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""EX_MOVE_DIFF"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""IMPL_VOL"", ""AAPL"")',
  '=RTD(""tos.rtd"", , ""Spreads"", ""AAPL"")',
  '=RTD(""tos.rtd"", 

Good. Now we have a list of lists. Ideally we don't want a list of lists, we just want a big list.

In [262]:
# Thanks ChatGPT for showing how to flatten a list.
mass_rtd_list = [item for sublist in rtd_call_per_ticker for item in sublist]

mass_rtd_list

['=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""LAST"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""LAST_SIZE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""BID_SIZE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""BID"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""ASK"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""ASK_SIZE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""NET_CHANGE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""PERCENT_CHANGE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""MARK_PERCENT_CHANGE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""PUT_CALL_RATIO"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""CALL_VOLUME_INDEX"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""PUT_VOLUME_INDEX"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""OPTION_VOLUME_INDEX"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""BACK_EX_MOVE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""FRONT_EX_MOvE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""EX_MOVE_DIFF"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""IMPL_VOL"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""Spreads"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""MRKT_MKR_MOVE"",

Bingo. Now we have to format it to match what we will actually paste into our program. (Found by Alex, will link to the 2019 repository. The author used Visual Basics to program this.)

In [263]:
# Format to replicate, with 'mass_rtd_list'
# XL_Cell_1.Formula = "=RTD(""tos.rtd"", , ""LAST"", ""QQQ"" )"
# XL_Cell_2.Formula = "=RTD(""tos.rtd"", , ""LAST"", ""AAPL"" )"
# XL_Cell_3.Formula = "=RTD(""tos.rtd"", , ""BID"", ""AAPL"" )"
# ' ...
# XL_Cell_LAST.Formula = "=RTD(""tos.rtd"", , ""INDICATOR"", ""TICKER"" )"
# XL_Cell_LAST.Formula = "=RTD(""tos.rtd"", , ""INDICATOR2"", ""TICKER2"" )"


# First thing to do is to fix the quotations. It gives ' ... we need  "
# DONE

In [264]:
# Now we add the XL_Cell_3 part.
# We are creating strings

string1 = "XL_Cell_1.Formula = " + "\""  + mass_rtd_list[1] + "\""

print(string1)

XL_Cell_1.Formula = "=RTD(""tos.rtd"", , ""LAST"", ""AAPL"")"


In [265]:
# Good. Now iterate and save into a final list.

final_rtd_calls = []
x = 0


# for x in range(len(mass_rtd_list)):
#     string1 = f"XL_Cell_{x}.Formula = " + "\""  + mass_rtd_list[x] + "\""
#     final_rtd_calls.append(string1)
#     x+1

# Make a quote to easily dump in the following...
q = "\""

for x in range(len(mass_rtd_list)):
    string1 = f"XL_Cell_{x}.Formula = {q}{mass_rtd_list[x]}{q}"
    final_rtd_calls.append(string1)
    x+1



final_rtd_calls

['XL_Cell_0.Formula = "=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")"',
 'XL_Cell_1.Formula = "=RTD(""tos.rtd"", , ""LAST"", ""AAPL"")"',
 'XL_Cell_2.Formula = "=RTD(""tos.rtd"", , ""LAST_SIZE"", ""AAPL"")"',
 'XL_Cell_3.Formula = "=RTD(""tos.rtd"", , ""BID_SIZE"", ""AAPL"")"',
 'XL_Cell_4.Formula = "=RTD(""tos.rtd"", , ""BID"", ""AAPL"")"',
 'XL_Cell_5.Formula = "=RTD(""tos.rtd"", , ""ASK"", ""AAPL"")"',
 'XL_Cell_6.Formula = "=RTD(""tos.rtd"", , ""ASK_SIZE"", ""AAPL"")"',
 'XL_Cell_7.Formula = "=RTD(""tos.rtd"", , ""NET_CHANGE"", ""AAPL"")"',
 'XL_Cell_8.Formula = "=RTD(""tos.rtd"", , ""PERCENT_CHANGE"", ""AAPL"")"',
 'XL_Cell_9.Formula = "=RTD(""tos.rtd"", , ""MARK_PERCENT_CHANGE"", ""AAPL"")"',
 'XL_Cell_10.Formula = "=RTD(""tos.rtd"", , ""PUT_CALL_RATIO"", ""AAPL"")"',
 'XL_Cell_11.Formula = "=RTD(""tos.rtd"", , ""CALL_VOLUME_INDEX"", ""AAPL"")"',
 'XL_Cell_12.Formula = "=RTD(""tos.rtd"", , ""PUT_VOLUME_INDEX"", ""AAPL"")"',
 'XL_Cell_13.Formula = "=RTD(""tos.rtd"", , ""OPTION_VOLUME_IN

In [266]:
print(final_rtd_calls[10])
print(final_rtd_calls[50])
print(final_rtd_calls[100])

XL_Cell_10.Formula = "=RTD(""tos.rtd"", , ""PUT_CALL_RATIO"", ""AAPL"")"
XL_Cell_50.Formula = "=RTD(""tos.rtd"", , ""ASK_SIZE"", ""GOOG"")"
XL_Cell_100.Formula = "=RTD(""tos.rtd"", , ""PUT_VOLUME_INDEX"", ""AMZN"")"


Bingo. Now we can copy and paste the output into our program.

In [267]:
# # Save list into a .txt file
# with open('rtd_calls.txt', 'w') as file:
#     for item in final_rtd_calls:
#         file.write(f"{item}\n")

In [280]:
print(f"Number of RTD Calls: {len(final_rtd_calls)}")

Number of RTD Calls: 11462


11,000 columns seems a bit ambitious. I think my computer will crash... So we will try to do that many, then scale down from there... After all, if I can iterate with 10, I should be able to iterate with 11462, and vice versa.

In [299]:
# # We'll keep going. Next is: 
# 'Load values returned from RTDServer
# Data_Col_1 = CStr(XL_Cell_1.Value)
# Data_Col_2 = CStr(XL_Cell_2.Value)
# ' ...
# Data_Col_LAST = CStr(XL_Cell_LAST.Value)

cstr_list = []

# For all formulas that we have, make "Data_Col_LAST = CStr(XL_Cell_LAST.Value)"
for x in range(len(final_rtd_calls)):
    string1 = f"Data_Col_{x} = CStr(XL_Cell_{x}.Value)"
    cstr_list.append(string1)


print(cstr_list[-3])
print(cstr_list[-2])
print(cstr_list[-1])


# Save list into a .txt file
with open('RTD_cstr_list.txt', 'w') as file:
    for item in cstr_list:
        file.write(f"{item}\n")  

Data_Col_11459 = CStr(XL_Cell_11459.Value)
Data_Col_11460 = CStr(XL_Cell_11460.Value)
Data_Col_11461 = CStr(XL_Cell_11461.Value)


In [269]:
# Now:
# ' Defining Excel source cells
# Private XL_Cell_1, XL_Cell_2, XL_Cell_3, XL_Cell_4, XL_Cell_5 As Excel.Range
# Private XL_Cell_6, XL_Cell_7, XL_Cell_8, XL_Cell_9, XL_Cell_10 As Excel.Range
# ' ...
# Private XL_Cell_6, XL_Cell_7, XL_Cell_8, XL_Cell_9, XL_Cell_LAST As Excel.Range

# We are going to try something different. Not sure why he is grouping by 5...

In [270]:
# Let's just iterate the "XL_Cell_1, XL_Cell_2, XL_Cell_3..."
# We will add "Private" and "As Excel.Range" later
xl_cell_list = []

# For all formulas that we have, make "XL_Cell_#"
for x in range(len(final_rtd_calls)):
    string1 = f"XL_Cell_{x}, "
    xl_cell_list.append(string1)

In [271]:
print(xl_cell_list[0])
print(xl_cell_list[10])
print(xl_cell_list[100])

XL_Cell_0, 
XL_Cell_10, 
XL_Cell_100, 


In [285]:
# Save list into a .txt file
with open('xl_cell_x.txt', 'w') as file:
    for item in xl_cell_list:
        file.write(f"{item} ")

In [272]:
# Now we have to:
# ' Define DB output columns
# Private Data_Col_1, Data_Col_2, Data_Col_3, Data_Col_4, Data_Col_5 As String
# Private Data_Col_6, Data_Col_7, Data_Col_8, Data_Col_9, Data_Col_10 As String
# ' ....

# Pretty much the same as before
# We will add "Private" and "As String" later
db_output_list = []

# For all formulas that we have, make "Data_Col_#"
for x in range(len(final_rtd_calls)):
    string1 = f"Data_Col_{x}, "
    db_output_list.append(string1)

In [273]:
print(db_output_list[0])
print(db_output_list[10])
print(db_output_list[100])


Data_Col_0, 
Data_Col_10, 
Data_Col_100, 


In [286]:
# Save list into a .txt file
with open('data_col_x.txt', 'w') as file:
    for item in db_output_list:
        file.write(f"{item} ")

In [290]:
# Now this:
# ' Define Cell Ranges
# XL_Cell_1 = oSheet.Cells(1, 1)
# XL_Cell_2 = oSheet.Cells(2, 1)
# XL_Cell_3 = oSheet.Cells(3, 1)
# ' ...
# XL_Cell_LAST = oSheet.Cells(LAST, 1)


cell_range_list = []

# For all formulas that we have, make "Data_Col_#"
for x in range(len(final_rtd_calls)):
    string1 = f"XL_Cell_{x} = oSheet.Cells({x}, 1)"
    cell_range_list.append(string1)


print(cell_range_list[-3])
print(cell_range_list[-2])
print(cell_range_list[-1])


# Save list into a .txt file
with open('osheet_cells.txt', 'w') as file:
    for item in cell_range_list:
        file.write(f"{item}\n")  

XL_Cell_11459 = oSheet.Cells(11459, 1)
XL_Cell_11460 = oSheet.Cells(11460, 1)
XL_Cell_11461 = oSheet.Cells(11461, 1)


In [276]:
# Now this: 
# ' column headers written at file creation
# Private strCSV_Header As String = "Date" & ", " & "/ES" & "AAPL Last" & "," &...& "AAPL Volume" & "," & "QQQ Last" &.& "lastTICKER + indicator"



In [297]:
# Now this: 
# ' Building data string
# strCSV_Data = CStr(tsTimeStamp) & ", " & Data_Col_1 & ", " & Data_Col_2 & ", " & Data_Col_3 & ... & ", " & Data_Col_LAST

str_csv_list = []

# string1 = " " & Data_Col_1 & ""


for x in range(len(final_rtd_calls)):
    string1 = f"& \", \" & Data_Col_{x}"
    str_csv_list.append(string1)


print(str_csv_list[-3])
print(str_csv_list[-2])
print(str_csv_list[-1])

# Save list into a .txt file
with open('RTD_str_csv_data.txt', 'w') as file:
    for item in str_csv_list:
        file.write(f"{item} ")  

& ", " & Data_Col_11459
& ", " & Data_Col_11460
& ", " & Data_Col_11461


I need to get one more list of the 11000+ elements. It is the names of the indicator + ticker...

In [320]:
# We will set the indicator + ticker as our column headers.
# ' column headers written at file creation
# Private strCSV_Header As String = "Date" & ", " & "/ES" & "AAPL Last" & "," &...& "AAPL Volume" & "," & "QQQ Last" &...& "lastTICKER + indicator"

# From each element on the mass_rtd_list, we need to capture ""AAPL"" + ""MARK"" to get the title
# titles should be: "AAPL MARK", and "AAPL PUT_CALL_RATIO"...

# To do so, extract the info from 'mass_rtd_list'
# We need to grab, ""MARK"" and ""AAPL"", for example.
mass_rtd_list[:3]

['=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""LAST"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""LAST_SIZE"", ""AAPL"")']

In [322]:
# Let's break down the first row.
# Our final output should be "AAPL MARK"
text = mass_rtd_list[0]
text

'=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")'

In [323]:
components = text.split(',')
components

['=RTD(""tos.rtd""', ' ', ' ""MARK""', ' ""AAPL"")']

In [324]:
print(components[0])
print(components[1])
print(components[2])
print(components[3])

=RTD(""tos.rtd""
 
 ""MARK""
 ""AAPL"")


In [327]:
indicator = components[2]

ticker = components[3]

print(indicator)
print(ticker)

 ""MARK""
 ""AAPL"")


In [328]:
indicator_components = indicator.split("\"")
ticker_components = ticker.split("\"")

print(indicator_components)
print(ticker_components)

[' ', '', 'MARK', '', '']
[' ', '', 'AAPL', '', ')']


In [329]:
header = ticker_components[2]  + " " + indicator_components[2]
header

'AAPL MARK'

Good. Now we iterate for the 11462 calls where we want to pull data from TOS. 

In [335]:
print(len(mass_rtd_list))
mass_rtd_list[:10]

11462


['=RTD(""tos.rtd"", , ""MARK"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""LAST"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""LAST_SIZE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""BID_SIZE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""BID"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""ASK"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""ASK_SIZE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""NET_CHANGE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""PERCENT_CHANGE"", ""AAPL"")',
 '=RTD(""tos.rtd"", , ""MARK_PERCENT_CHANGE"", ""AAPL"")']

In [348]:
# Flow of breaking down string:
text = mass_rtd_list[1]

components = text.split(',')

indicator = components[2]
ticker = components[3]

indicator_components = indicator.split("\"")
ticker_components = ticker.split("\"")

header = ticker_components[2]  + " " + indicator_components[2]
header

'AAPL LAST'

In [349]:
# Reminder of our target:
# "/ES" & ", " & "/NQ" & ", " & "/RTY" & ", " &
# We have 'AAPL LAST', we need: '"AAPL LAST" & ", " & '
q = "\""
header = f"{q}{header}{q} & {q}, {q} & "

header

# We'll add this to the following

'"AAPL LAST" & ", " & '

In [352]:
# Iterate the flow
header_list = []
q = "\""
x = 0

for x in range(len(mass_rtd_list)):
    text = mass_rtd_list[x]
    components = text.split(',')
    indicator = components[2]
    ticker = components[3]
    indicator_components = indicator.split("\"")
    ticker_components = ticker.split("\"")
    header = ticker_components[2]  + " " + indicator_components[2]
    header = f"{q}{header}{q} & {q}, {q} &"
    header_list.append(header)


header_list[:10]

['"AAPL MARK" & ", " &',
 '"AAPL LAST" & ", " &',
 '"AAPL LAST_SIZE" & ", " &',
 '"AAPL BID_SIZE" & ", " &',
 '"AAPL BID" & ", " &',
 '"AAPL ASK" & ", " &',
 '"AAPL ASK_SIZE" & ", " &',
 '"AAPL NET_CHANGE" & ", " &',
 '"AAPL PERCENT_CHANGE" & ", " &',
 '"AAPL MARK_PERCENT_CHANGE" & ", " &']

In [353]:
# Save list into a .txt file
with open('RTD_header_list.txt', 'w') as file:
    for item in header_list:
        file.write(f"{item} ")  